File: rust/skialin-core/src/runtime_effect.rs
use crate::{sys, ColorFilter, Matrix, Shader};

pub struct RuntimeEffect(pub(crate) *mut sys::SkRuntimeEffect);

impl RuntimeEffect {
    /// Compiles an SkSL shader effect. The SkSL must define
    /// `vec4 main(vec2 coord) { ... }` returning a premultiplied color.
    /// On failure, returns the compiler's error text.
    pub fn make_for_shader(sksl: &str) -> Result<Self, String> {
        let mut error: *mut sys::SkData = std::ptr::null_mut();
        let effect = unsafe { sys::skialin_bridge_RuntimeEffect_MakeForShader(sksl.as_ptr().cast(), sksl.len(), &mut error) };
        Self::from_result(effect, error)
    }

    /// Compiles an SkSL color-filter effect. The SkSL must define
    /// `vec4 main(vec4 inColor) { ... }`. On failure, returns the
    /// compiler's error text.
    pub fn make_for_color_filter(sksl: &str) -> Result<Self, String> {
        let mut error: *mut sys::SkData = std::ptr::null_mut();
        let effect = unsafe { sys::skialin_bridge_RuntimeEffect_MakeForColorFilter(sksl.as_ptr().cast(), sksl.len(), &mut error) };
        Self::from_result(effect, error)
    }

    fn from_result(effect: *mut sys::SkRuntimeEffect, error: *mut sys::SkData) -> Result<Self, String> {
        if !effect.is_null() {
            return Ok(RuntimeEffect(effect));
        }
        let message = unsafe { crate::Data::from_raw(error) }.map(|d| String::from_utf8_lossy(d.as_bytes()).into_owned()).unwrap_or_default();
        Err(message)
    }

    /// `uniforms` is a raw byte buffer packed to match the SkSL uniform
    /// block's layout (the caller is responsible for knowing that layout).
    pub fn make_shader(&self, uniforms: &[u8], children: &[&Shader], local_matrix: Option<&Matrix>) -> Option<Shader> {
        let child_ptrs: Vec<*mut sys::SkShader> = children.iter().map(|c| c.0).collect();
        let matrix_ptr = local_matrix.map_or(std::ptr::null(), |m| &m.0);
        unsafe {
            Shader::from_raw(sys::skialin_bridge_RuntimeEffect_makeShader(
                self.0,
                uniforms.as_ptr(),
                uniforms.len(),
                child_ptrs.as_ptr(),
                child_ptrs.len(),
                matrix_ptr,
            ))
        }
    }

    /// `uniforms` is a raw byte buffer packed to match the SkSL uniform
    /// block's layout (the caller is responsible for knowing that layout).
    pub fn make_color_filter(&self, uniforms: &[u8], children: &[&ColorFilter]) -> Option<ColorFilter> {
        let child_ptrs: Vec<*mut sys::SkColorFilter> = children.iter().map(|c| c.0).collect();
        unsafe { ColorFilter::from_raw(sys::skialin_bridge_RuntimeEffect_makeColorFilter(self.0, uniforms.as_ptr(), uniforms.len(), child_ptrs.as_ptr(), child_ptrs.len())) }
    }
}

impl Drop for RuntimeEffect {
    fn drop(&mut self) {
        unsafe { sys::skialin_bridge_RuntimeEffect_unref(self.0) };
    }
}

File: src/main/kotlin/org/skialin/RuntimeEffect.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

/**
 * A compiled SkSL shader or color-filter effect. Mirrors Skia's `SkRuntimeEffect`.
 *
 * @throws IllegalArgumentException if [sksl] fails to compile.
 */
class RuntimeEffect private constructor(
    ptr: Long,
) : Managed(ptr, RuntimeEffectNative::nRelease) {
    /**
     * [uniforms] is a raw byte buffer packed to match the SkSL uniform block's layout
     * (the caller is responsible for knowing that layout); pass `null` if the effect
     * declares no uniforms.
     */
    fun makeShader(
        uniforms: ByteArray? = null,
        children: Array<Shader> = emptyArray(),
        localMatrix: Matrix33? = null,
    ): Shader? {
        val childPtrs = LongArray(children.size) { children[it].nativePtr }
        val ptr = RuntimeEffectNative.nMakeShader(nativePtr, uniforms, childPtrs, localMatrix?.values)
        return if (ptr == 0L) null else Shader(ptr)
    }

    /**
     * [uniforms] is a raw byte buffer packed to match the SkSL uniform block's layout
     * (the caller is responsible for knowing that layout); pass `null` if the effect
     * declares no uniforms.
     */
    fun makeColorFilter(
        uniforms: ByteArray? = null,
        children: Array<ColorFilter> = emptyArray(),
    ): ColorFilter? {
        val childPtrs = LongArray(children.size) { children[it].nativePtr }
        val ptr = RuntimeEffectNative.nMakeColorFilter(nativePtr, uniforms, childPtrs)
        return if (ptr == 0L) null else ColorFilter(ptr)
    }

    companion object {
        /** [sksl] must define `vec4 main(vec2 coord) { ... }` returning a premultiplied color. */
        fun makeForShader(sksl: String): RuntimeEffect = RuntimeEffect(RuntimeEffectNative.nMakeForShader(sksl))

        /** [sksl] must define `vec4 main(vec4 inColor) { ... }`. */
        fun makeForColorFilter(sksl: String): RuntimeEffect = RuntimeEffect(RuntimeEffectNative.nMakeForColorFilter(sksl))
    }
}

private object RuntimeEffectNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nMakeForShader(sksl: String): Long

    external fun nMakeForColorFilter(sksl: String): Long

    external fun nRelease(ptr: Long)

    external fun nMakeShader(
        ptr: Long,
        uniforms: ByteArray?,
        children: LongArray,
        localMatrix: FloatArray?,
    ): Long

    external fun nMakeColorFilter(
        ptr: Long,
        uniforms: ByteArray?,
        children: LongArray,
    ): Long
}
